File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return head; // Handle empty or single-node list
        }
        
        stack<ListNode*> rev;
        
        ListNode *curr = head;
        while(curr != NULL) {
            rev.push(curr);
            curr = curr->next;
        }
        
        ListNode* head1 = rev.top();
        ListNode* revHead = rev.top();
        rev.pop();
        
        while(!rev.empty()){
            revHead->next = rev.top();
            revHead = revHead->next;
            rev.pop();
        }
        
        // Terminate the reversed list properly
        revHead->next = nullptr;
        
        return head1;
    }
};

File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head; // Handle empty or single-node list
        }
        
        Stack<ListNode> rev = new Stack<>();
        
        ListNode curr = head;
        while (curr != null) {
            rev.push(curr);
            curr = curr.next;
        }
        
        ListNode head1 = rev.peek();
        ListNode revHead = rev.peek();
        rev.pop();
        
        while (!rev.empty()) {
            revHead.next = rev.peek();
            revHead = revHead.next;
            rev.pop();
        }
        
        // Terminate the reversed list properly
        revHead.next = null;
        
        return head1;
    }
}